Allow doing start-end RANGE requests against blobs

thundergolfer opened the pull request:
Addresses https://github.com/oras-project/rust-oci-client/issues/162

flavio reviewed (commented):
Seems fine, can you address the build failure?

Moreover, is there some way we can test this feature inside of our CI?

thundergolfer commented:
Oh damn sorry I thought this was compiling because I was pointing at my fork but I had a cached `main` 😖

thundergolfer commented:
@flavio I updated the implementation and added a test. During reimplementation I switch to using `pull_blob_stream_partial` and ran into an issue where `VerifyingStream` doesn't work when you get a partial response back (the digests will never match). 

I'm not sure how this was working before? Let me know what you think :)

thundergolfer commented on src/client.rs:
```diff
@@ -1142,14 +1146,19 @@ impl Client {
+            )?)),
             StatusCode::PARTIAL_CONTENT => Ok(BlobResponse::Partial(stream_from_response(
-                response, &layer,
+                response, &layer, false,
```
If the `offset` was ever non-zero this method should already have been failing on stream verification? I checked registry spec and couldn't see anything about returned specialized digests for partial reads.

thomastaylor312 replied:
It makes sense that partial pulls can't be validated, so we should probably document the `pull_blob_stream_partial` that partial pulls by nature cannot be validated and as such should be validated by the caller when complete

thomastaylor312 replied:
Actually, now that I'm looking at this more, we might need to return the two possible digests to validate against to the caller on a partial request, otherwise they won't have access to the digest of the header (probably as new data members of `SizedStream`). What do other clients do/expect to handle digest validation in these cases?

thomastaylor312 replied:
@thundergolfer If you'd like/are willing to do this, that would be great! If not, let me know and I'll switch this to an approval and I'll change it after merge since it is a preexisting edge case we didn't account for originally

thundergolfer replied:
Hey @thomastaylor312, I've switched focus off this project so I'll be unlikely to get to it soon.

thomastaylor312 replied:
Ok sounds good. I'll merge this as is and fix up. Thanks for adding this!

flavio commented:
@thundergolfer thanks for having introduced the test case. 

@thomastaylor312 can you take a quick look too?

thundergolfer commented:
Nice, thanks. Let me know if you need anything else before merge 👌

thomastaylor312 commented:
Sorry forgot to add context when I hit the submit button. This all looks good to me. One minor point around adding some clarification to the docs for a function around the security implications
